### ISSUE TITLE
EFDT debug_one sometimes doesn't work

### ISSUE THREAD

**Issue by FedericoMz** [online-ml/river#1248]
### Discussed in https://github.com/online-ml/river/discussions/1247

<div type='discussions-op-text'>

<sup>Originally posted by **FedericoMz** May 29, 2023</sup>
Hello! I've been experimenting with Extremely Fast Decision Tree. 

I have a stream of records and labels. After training my model on each record and label, I use the _debug_one_ method on the previously-labeled records to see if the decision rule had changed. Unfortunately, there is always a problematic record which gives my this error.

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[32], line 1
----> 1 model.debug_one(x_past)

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/hoeffding_tree.py:364, in HoeffdingTree.debug_one(self, x)
    361         except KeyError:
    362             child_index, _ = node.most_common_path()
--> 364         _print(node.repr_branch(child_index))
    366 return buffer.getvalue()

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/nodes/branch.py:195, in NominalMultiwayBranch.repr_branch(self, index, shorten)
    194 def repr_branch(self, index: int, shorten=False):
--> 195     feat_val = self._r_mapping[index]
    197     if shorten:
    198         return str(feat_val)

KeyError: 8
```

It's worth noting that both _predict_one_ and _predict_proba_one_ work. However, the issue seems to affect _draw_ too.

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[26], line 1
----> 1 model.draw()

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/hoeffding_tree.py:477, in HoeffdingTree.draw(self, max_depth)
    471     dot.node(f"{child_no}", text, fillcolor=fillcolor, style="filled")
    473     if parent_no is not None:
    474         dot.edge(
    475             f"{parent_no}",
    476             f"{child_no}",
--> 477             xlabel=parent.repr_branch(branch_index, shorten=True),
    478         )
    480 return dot

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/nodes/branch.py:195, in NominalMultiwayBranch.repr_branch(self, index, shorten)
    194 def repr_branch(self, index: int, shorten=False):
--> 195     feat_val = self._r_mapping[index]
    197     if shorten:
    198         return str(feat_val)

KeyError: 6
```

If I train the model a second time on the problematic record, I can use _debug_one_, but the issue persists with _draw_.

I don't think the problem lies with the data, because:
1) The model learns from the problematic record, and _debug_one_ works as aspected. The problem arises after training the model on some new records;
2) I tried multiple set of data and there's always a (different) problematic record.</div>

**Comment by MaxHalford** [comment]
Is your data sparse? I think what's going on is that some of your samples are incomplete. So the draw methods tries to access a feature to determine a split, and that feature isn't available.

**Comment by smastelini** [comment]
Hi @FedericoMz, more details about your data would be nice. I am guessing you are using categorical features, am I correct?

**Comment by smastelini** [comment]
Hi @FedericoMz, any updates on your issue?

**Comment by MaxHalford** [comment]
Closing because there's no activity. But feel free to reopen if necessary.

**Comment by FedericoMz** [comment]
Hello

I was sure I had posted the notebook with the reproducible issue, but I must have forgotten about it. You can find it [here](https://github.com/FedericoMz/RiverTest). I retested with the latest version of River, and the problem is still there.

In the second cell, an EFDT model is trailed without ever asking for explanations (neither prediction probabilities nor decision rules). This is optional, but it speeds up this reproducible example.

In the third cell, whenever the model is trained with a new record, we check the predict proba and decision rules of past records. At some point (5% of this second dataset), running _both_ predict_proba_one and debug_one leads to crashing. The order doesn't matter.

**Comment by MaxHalford** [comment]
@smastelini could you take a look? :)

**Comment by smastelini** [comment]
Yeap! I'll check it out

**Comment by smastelini** [comment]
It took me a while, but I found the problem. You can check the fix in #1337.

@FedericoMz, can you please confirm whether the problem is solved?

**Comment by MaxHalford** [comment]
Good job Saulo!
